Initial rpmbuld failure - test_disable_ip_forwarding

During the initial review of #2712 (comment https://github.com/rockstor/rockstor-core/pull/2712#pullrequestreview-1687632231) there was an as yet unexplained fail in a new test regarding ip forwarding:
```
======================================================================
FAIL: test_disable_ip_forwarding (rockstor.system.tests.test_system_network.SystemNetworkTests)
test proper call of os.remove() and final sysctl command
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/buildbot/worker/Poetry-Build-on-Leap15-4/rpmbuild/rockstor-core-5.0.5-0/src/rockstor/system/tests/test_system_network.py", line 795, in test_disable_ip_forwarding
    self.mock_os_remove.assert_called_once_with("/etc/sysctl.d/99-tailscale.conf")
  File "/usr/lib64/python3.9/unittest/mock.py", line 918, in assert_called_once_with
    raise AssertionError(msg)
AssertionError: Expected 'remove' to be called once. Called 0 times.
----------------------------------------------------------------------
Ran 279 tests in 32.169s
FAILED (failures=1)
```
On further fresh rpmbuild attempts (other OS/arch targets) this has been seen again on all systems that have not previously done an rpmbuild with this test included. We likely have some inadvertent system sensitivity/interaction here.

Apologies for failing with these tests... It seems I missed some mocking somewhere. Here's what I noticed:

I ran the test once first and it all passed, even after stopping, uninstall tailscale and its repo (and after a reboot):
```
buildvm155:/opt/rockstor/src/rockstor # poetry run django-admin test -v 2 -p test_system_network.py
Skipping setup of unused database(s): default, smart_manager.
System check identified no issues (0 silenced).
test_disable_ip_forwarding (rockstor.system.tests.test_system_network.SystemNetworkTests)
test proper call of os.remove() and final sysctl command ... ok
test_enable_ip_forwarding_syctl (rockstor.system.tests.test_system_network.SystemNetworkTests)
enable_ip_forwarding() triggers sysctl config reload ... ok
test_enable_ip_forwarding_write (rockstor.system.tests.test_system_network.SystemNetworkTests)
enable_ip_forwarding() calls write with the correct contents ... ok
test_get_con_config (rockstor.system.tests.test_system_network.SystemNetworkTests)
This tests for correct parsing of nmcli connection config by get_con_config(), ... ok
test_get_con_config_con_not_found (rockstor.system.tests.test_system_network.SystemNetworkTests)
Test get_con_config() if connection is not found / vanished. ... ok
test_get_con_config_exception (rockstor.system.tests.test_system_network.SystemNetworkTests)
Test get_con_config() if nmcli returns error code != 10. ... ok
test_get_dev_config (rockstor.system.tests.test_system_network.SystemNetworkTests)
This tests for correct parsing of nmcli device config by get_dev_config(), ... ok
test_get_dev_config_dev_not_found (rockstor.system.tests.test_system_network.SystemNetworkTests)
Test get_dev_config() if device is not found / vanished. ... ok
test_get_dev_config_exception (rockstor.system.tests.test_system_network.SystemNetworkTests)
Test get_dev_config() if nmcli returns error code != 10. ... ok

----------------------------------------------------------------------
Ran 9 tests in 0.014s

OK
```

I noticed, however, that I had the file `/etc/sysctl.d/99-tailscale.conf` physically present on the system, hence why it worked. Indeed, that failing `test_disable_ip_forwarding()` test checks if that file exists and tries to verify `os.remove()` was called on it.
```
buildvm155:/opt/rockstor/src/rockstor # ls -lah /etc/sysctl.d/
total 8.0K
drwxr-xr-x 1 root root   58 Oct 17 18:51 .
drwxr-xr-x 1 root root 3.6K Oct 19 15:11 ..
-rw-r--r-- 1 root root   35 Aug 11 10:49 70-yast.conf
-rw-r--r-- 1 root root   57 Oct 19 15:13 99-tailscale.conf
```

Now, if I manually delete that file and run the tests again, I get the same failure you have:
```
buildvm155:/opt/rockstor/src/rockstor # ls -lah /etc/sysctl.d/
total 4.0K
drwxr-xr-x 1 root root   24 Oct 19 15:15 .
drwxr-xr-x 1 root root 3.6K Oct 19 15:11 ..
-rw-r--r-- 1 root root   35 Aug 11 10:49 70-yast.conf


buildvm155:/opt/rockstor/src/rockstor # poetry run django-admin test -v 2 -p test_system_network.py
Skipping setup of unused database(s): default, smart_manager.
System check identified no issues (0 silenced).
test_disable_ip_forwarding (rockstor.system.tests.test_system_network.SystemNetworkTests)
test proper call of os.remove() and final sysctl command ... FAIL
test_enable_ip_forwarding_syctl (rockstor.system.tests.test_system_network.SystemNetworkTests)
enable_ip_forwarding() triggers sysctl config reload ... ok
test_enable_ip_forwarding_write (rockstor.system.tests.test_system_network.SystemNetworkTests)
enable_ip_forwarding() calls write with the correct contents ... ok
test_get_con_config (rockstor.system.tests.test_system_network.SystemNetworkTests)
This tests for correct parsing of nmcli connection config by get_con_config(), ... ok
test_get_con_config_con_not_found (rockstor.system.tests.test_system_network.SystemNetworkTests)
Test get_con_config() if connection is not found / vanished. ... ok
test_get_con_config_exception (rockstor.system.tests.test_system_network.SystemNetworkTests)
Test get_con_config() if nmcli returns error code != 10. ... ok
test_get_dev_config (rockstor.system.tests.test_system_network.SystemNetworkTests)
This tests for correct parsing of nmcli device config by get_dev_config(), ... ok
test_get_dev_config_dev_not_found (rockstor.system.tests.test_system_network.SystemNetworkTests)
Test get_dev_config() if device is not found / vanished. ... ok
test_get_dev_config_exception (rockstor.system.tests.test_system_network.SystemNetworkTests)
Test get_dev_config() if nmcli returns error code != 10. ... ok

======================================================================
FAIL: test_disable_ip_forwarding (rockstor.system.tests.test_system_network.SystemNetworkTests)
test proper call of os.remove() and final sysctl command
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/rockstor/src/rockstor/system/tests/test_system_network.py", line 795, in test_disable_ip_forwarding
    self.mock_os_remove.assert_called_once_with("/etc/sysctl.d/99-tailscale.conf")
  File "/usr/lib64/python3.9/unittest/mock.py", line 918, in assert_called_once_with
    raise AssertionError(msg)
AssertionError: Expected 'remove' to be called once. Called 0 times.

----------------------------------------------------------------------
Ran 9 tests in 0.014s

FAILED (failures=1)
```

The devil is in the details here: the order in which the tests are run: `test_disable_ip_forwarding()` is run before `test_enable_ip_forwarding_write()` and `test_enable_ip_forwarding_sysctl()`. The latter two indeed are testing if that `99-tailscale.conf` file is written and then sysctl called to refresh its config from that file. It seems I failed to properly mock that writing as the file is *actually* written to disk after this round of tests with `disable_ip_forwarding()` that failed:
```
buildvm155:/opt/rockstor/src/rockstor # ls -lah /etc/sysctl.d/
total 8.0K
drwxr-xr-x 1 root root   58 Oct 19 15:15 .
drwxr-xr-x 1 root root 3.6K Oct 19 15:11 ..
-rw-r--r-- 1 root root   35 Aug 11 10:49 70-yast.conf
-rw-r--r-- 1 root root   57 Oct 19 15:15 99-tailscale.conf
```

If we run again these tests, they all pass now given the file is present:
```
buildvm155:/opt/rockstor/src/rockstor # poetry run django-admin test -v 2 -p test_system_network.py
Skipping setup of unused database(s): default, smart_manager.
System check identified no issues (0 silenced).
test_disable_ip_forwarding (rockstor.system.tests.test_system_network.SystemNetworkTests)
test proper call of os.remove() and final sysctl command ... ok
test_enable_ip_forwarding_syctl (rockstor.system.tests.test_system_network.SystemNetworkTests)
enable_ip_forwarding() triggers sysctl config reload ... ok
test_enable_ip_forwarding_write (rockstor.system.tests.test_system_network.SystemNetworkTests)
enable_ip_forwarding() calls write with the correct contents ... ok
test_get_con_config (rockstor.system.tests.test_system_network.SystemNetworkTests)
This tests for correct parsing of nmcli connection config by get_con_config(), ... ok
test_get_con_config_con_not_found (rockstor.system.tests.test_system_network.SystemNetworkTests)
Test get_con_config() if connection is not found / vanished. ... ok
test_get_con_config_exception (rockstor.system.tests.test_system_network.SystemNetworkTests)
Test get_con_config() if nmcli returns error code != 10. ... ok
test_get_dev_config (rockstor.system.tests.test_system_network.SystemNetworkTests)
This tests for correct parsing of nmcli device config by get_dev_config(), ... ok
test_get_dev_config_dev_not_found (rockstor.system.tests.test_system_network.SystemNetworkTests)
Test get_dev_config() if device is not found / vanished. ... ok
test_get_dev_config_exception (rockstor.system.tests.test_system_network.SystemNetworkTests)
Test get_dev_config() if nmcli returns error code != 10. ... ok

----------------------------------------------------------------------
Ran 9 tests in 0.013s

OK
```

--- a/src/rockstor/system/tests/test_system_network.py
+++ b/src/rockstor/system/tests/test_system_network.py
@@ -13,7 +13,7 @@ You should have received a copy of the GNU General Public License
 along with this program. If not, see <https://www.gnu.org/licenses/>.
 """
 import unittest
-from unittest.mock import patch, mock_open, call, MagicMock
+from unittest.mock import patch, mock_open, call
 
 from system.exceptions import CommandException
 from system.network import (
@@ -24,6 +24,25 @@ from system.network import (
     SYSCTL,
     disable_ip_forwarding,
 )
+
+
+class MockDirEntry:
+    """Create a fake DirEntry object
+    A DirEntry object is what's os.scandir returns. To properly mock os.scandir,
+    we thus need to re-create what it would return. We currently only need some
+    of the attributes that a real DirEntry has, so let's include only these:
+    - name
+    - path
+    - is_file: bool
+    """
+
+    def __init__(self, name, path, is_file):
+        self.name = name
+        self.path = path
+        self._is_file = is_file
+
+    def is_file(self):
+        return self._is_file
 
 
 class SystemNetworkTests(unittest.TestCase):
@@ -741,7 +760,7 @@ class SystemNetworkTests(unittest.TestCase):
         with self.assertRaises(CommandException):
             get_con_config(con_name)
 
-    def test_enable_ip_forwarding_write(self):
+    def test_enable_ip_forwarding(self):
         """enable_ip_forwarding() calls write with the correct contents
         Enable_ip_forwarding should write to /etc/sysctl.d/99-tailscale.conf
         with the following lines
@@ -765,17 +784,7 @@ class SystemNetworkTests(unittest.TestCase):
             ]
             m().write.assert_has_calls(calls, any_order=False)
 
-    def test_enable_ip_forwarding_syctl(self):
-        """enable_ip_forwarding() triggers sysctl config reload
-        Sysctl configuration should be reloaded for the contents of
-        the /etc/sysctl.d/99-tailscale.conf file.
-        """
-        priority = 99
-        name = "tailscale"
-        file_path = f"{SYSCTL_CONFD_PATH}{priority}-{name}.conf"
-
         # Test that run_command was called as expected
-        enable_ip_forwarding(name=name, priority=priority)
         self.mock_run_command.assert_called_once_with(
             [SYSCTL, "-p", file_path], log=True
         )
@@ -783,8 +792,18 @@ class SystemNetworkTests(unittest.TestCase):
     def test_disable_ip_forwarding(self):
         """test proper call of os.remove() and final sysctl command"""
         # mock os.scandir()
-        self.mock_os_scandir = MagicMock()
-        self.mock_os_scandir.return_value = ["70-yast.conf", "99-tailscale.conf"]
+        self.patch_os_scandir = patch("system.network.os.scandir")
+        self.mock_os_scandir = self.patch_os_scandir.start()
+        self.mock_os_scandir.return_value.__enter__.return_value = [
+            MockDirEntry(
+                name="99-tailscale.conf",
+                path="/etc/sysctl.d/99-tailscale.conf",
+                is_file=True,
+            ),
+            MockDirEntry(
+                name="70-yast.conf", path="/etc/sysctl.d/70-yast.conf", is_file=True
+            ),
+        ]
         # mock os.remove()
         self.patch_os_remove = patch("system.network.os.remove")
         self.mock_os_remove = self.patch_os_remove.start()

